pkg/virt-controller/watch/ovm.go

```diff
@@ -0,0 +1,184 @@
+			controller.addToQueue(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {},
+	})
+
+	return controller
+}
+
+// addToQueue takes an object reported by handler, lookup its namespace and key
+// test wheter it is not deleted - if so ignore and if not add it for processing
+func (c *OVMController) addToQueue(obj interface{}) {
+	var (
+		key string
+		err error
+	)
+
+	if key, err = cache.DeletionHandlingMetaNamespaceKeyFunc(obj); err != nil {
+		// The error happened but it is not critical.
+		// Let kubernetes handle and continue.
+		runtime.HandleError(err)
+		return
+	}
+
+	c.workqueue.AddRateLimited(key)
+}
+
+// Run starts the controller. It setups the caches and waits for them to sync.
+// It spaws all workers and blocks until the stopCh closes.
+func (c *OVMController) Run(workers int, stopCh <-chan struct{}) error {
+	// Not defering the CrashHandling. It should be safe for the controller
+	// to crash, be restarted and continue work
+	//defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	log.Print("Starting the OVM controller")
+
+	log.Print("Syncing cache")
+	if ok := cache.WaitForCacheSync(stopCh, c.ovmInformer.HasSynced); !ok {
+		return fmt.Errorf("Caches cannot be synced")
+	}
+
+	log.Print("Spawning OVM controller workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	log.Print("Workers started")
+	<-stopCh
+	log.Print("Stopped OVM controller")
+
+	return nil
+}
+
+// worker is an infinite loop. Takes one item from queue and process it.
+// Terminates when the queue is marker for shutdown.
+func (c *OVMController) worker() {
+	for c.processNextItem() {
+	}
+}
+
+// processNextItem takes one item from the workqueue and process it.
+// it return false when the queue is marker for shutdown
+func (c *OVMController) processNextItem() bool {
+
+	obj, quit := c.workqueue.Get()
+	if quit {
+		return false
+	}
+	defer c.workqueue.Done(obj)
+
+	var (
+		key string
+		ok  bool
+	)
+
+	if key, ok = obj.(string); !ok {
+		// no string key, so forget it and handle the error in kubernetes
+		c.workqueue.Forget(obj)
+		runtime.HandleError(fmt.Errorf("Expected string key, but got %#v", obj))
+	}
+
+	if err := c.processItem(key); err != nil {
+		runtime.HandleError(err)
+	}
+
+	log.Printf("Processing done for key: %s", key)
+	c.workqueue.Forget(obj)
+
+	return true
+}
+
+// processItem takes the key, loads the item from cluster.
+func (c *OVMController) processItem(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("Cannot get right namespace and key for key: %s. Err: %s", key, err.Error())
+	}
+
+	_, err = c.getOvm(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Do some magic processing here at this point.
+
+	return nil
+}
+
+func (c *OVMController) getOvm(namespace, name string) (*kubeapiv1.OfflineVirtualMachine, error) {
+	obj, exists, err := c.ovmInformer.GetIndexer().GetByKey(namespace + "/" + name)
```

Comment:
That might work too, but normally GetStore().GetByKey() is used if you don't require access to a manually registered index.